### DISCUSSION TITLE
Dismissing more than one feature

### DISCUSSION THREAD

**Discussion by nmalevich** [pointfreeco/swift-composable-architecture#2895] [General]
### Description

When I have migrated to 1.9 version from 1.3 and started to use @ObservableState, @Presents, observe {}(I use UIKit for UI) and all what you recommend. All is good and thank you for your effort to improve TCA.

But now I have one problem with dismiss dependency. I didn't have this problem when I used 1.3 version. The problem is when I execute dismiss() from dismiss dependency at my child feature, not only my child feature set nil, but the parent feature too. The parent feature is added to main reducer by Scope(state:action:) and child feature by .ifLet(state: \.$destination, action: \.destination)

Yes, I have tried to use just callback effects from child features, detect them in parent reducer and set destination nil and it works well. But I want to use more simple way and exactly want to understand why this bug reproduce

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

After dismiss() set nil for only one(child) feature

### Actual behavior

After dismiss() set nil for two(parent and child) features

### Steps to reproduce

I can't share all my project, because it's not mine and I don't have this opportunity, but I can share with you some logs and results

I will share with you two variants of my app. And this variants have only one difference
1) With bug. I have this sending line on View that is connected with child feature
<img width="449" alt="Снимок экрана 2024-03-06 в 16 41 14" src="https://github.com/pointfreeco/swift-composable-architecture/assets/147712544/40d103c9-54d3-436e-aaa6-15254bfc8927">

2) No bug. I commented this sending line
<img width="470" alt="Снимок экрана 2024-03-06 в 16 42 11" src="https://github.com/pointfreeco/swift-composable-architecture/assets/147712544/0902943a-91e4-45e0-b5dc-77a814cefd7a">

This action "test" isn't detected on any reducers, so it can't influence for something

1) 
[test1.txt](https://github.com/pointfreeco/swift-composable-architecture/files/14510730/test1.txt)
<img width="640" alt="Снимок экрана 2024-03-06 в 16 48 07" src="https://github.com/pointfreeco/swift-composable-architecture/assets/147712544/db483008-ebf7-4c36-a4df-2f65f144eb90">
<img width="645" alt="Снимок экрана 2024-03-06 в 16 48 18" src="https://github.com/pointfreeco/swift-composable-architecture/assets/147712544/ef3eeaaf-d0cf-420b-b26a-7f551627850a">

2)
[test2.txt](https://github.com/pointfreeco/swift-composable-architecture/files/14510810/test2.txt)

You can also pay your attention to logs like this "<decode: bad range for [%@] got [offs:346 len:932 within:0]>". I'm sure that this error log is connected with the problem

### The Composable Architecture version information

1.9.0

### Destination operating system

iOS 17.3

### Xcode version information

15.2

### Swift Compiler version information

```shell
swift-driver version: 1.87.3 Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5)
Target: x86_64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @nmalevich, can you please provide a project that demonstrates the behavior you are seeing? I don't think this is an issue with the library so I am going to convert this to a discussion for now. We can re-open an issue if it turns out to be something wrong with TCA.

**Comment by nmalevich** [comment]
No, sorry, I can't provide my project. And yes, I can't say that this bug is only connected with TCA, I agree that can be my mistake. But for me it is really strange that one line which really wouldn't do anything, can influence somehow.

Maybe you can understand why sending effect that isn't detected wherever can be cause

**Reply by mbrandonw** [reply]
Hi @nmalevich, I'm not asking for your project. I'm asking for a new project that demonstrates the problem. Often the act of trying to create a small, reproducible project shows where things went wrong in the original code.

If you can provide such a small project we will happily look into it. But without that there isn't much we can do from just a short description and some screenshots of code.

**Reply by nmalevich** [reply]
Hi @mbrandonw, yes, I will try to do it soon. Now I have solved this problem by replacing dismiss() for callback effect and just setting nil to state.destination

**Reply by nmalevich** [reply]
Hi @mbrandonw

Finally I can share project with bad behavior of dismiss().
[DismissErrorExample.zip](https://github.com/pointfreeco/swift-composable-architecture/files/14946366/DismissErrorExample.zip)
You can see that we can't pop from third(red) screen

Please, feel free to look that.
Thanks

**Reply by mbrandonw** [reply]
HI @nmalevich, if you attach `_printChanges()` to the root reducer you see see that as far as TCA is concerned dismiss is working just fine:

```swift
received action:
  HomeFeature.Action.actionSubjects(
    .destination(
      .presented(
        .subject(
          .destination(
            .presented(
              .scan(
                .actionNavigationBar(
                  .callback(.didTapLeftItemButton)
                )
              )
            )
          )
        )
      )
    )
  )
  (No state changes)

received action:
  HomeFeature.Action.actionSubjects(
    .destination(
      .presented(
        .subject(
          .destination(.dismiss)
        )
      )
    )
  )
  HomeFeature.State(
    _stateNavigationBar: NavigationBarFeature.State(…),
    _stateSubjects: HomeSubjectsFeature.State(
      _destination: .subject(
        SubjectFeature.State(
-         _destination: .scan(
-           ScanFeature.State(
-             stateNavigationBar: NavigationBarFeature.State(
-               backgroundColor: UICachedDeviceWhiteColor(),
-               leftItemImage: UIImage(),
-               leftItemPadding: 9.0,
-               title: "Scan",
-               titleColor: #1 UICachedDeviceWhiteColor()
-             )
-           )
-         ),
+         _destination: nil,
          _stateNavigationBar: NavigationBarFeature.State(…),
          _typedTask: ""
        )
      )
    )
  )
```

This shows that `.didTapLeftItemButton` is sent, which causes `.dismiss` to be sent, which then `nil`s out the destination.

So any problem with navigation is due to SwiftUI not updating properly. There are lots of bugs in SwiftUI navigation, and it's what of the reasons why `NavigationStack` was introduced in iOS 16.

You can even put a print next to the `NavigationLink` to see that the view is re-computing and the state is correct:

```swift
.background {
  let _ = print("store.destination == nil", store.destination == nil)
  NavigationLink(…)
}
```

This will print:

```
store.destination == nil true
store.destination == nil false
store.destination == nil true
```

So, not much that can be done I'm afraid.

We have found that sometimes `@Environment(\.dismiss)` works around some SwiftUI bugs (though also can cause others). So if you use that in `NavigationBarView` instead of sending the `didTapLeftItemButton` button it looks like it works.

**Reply by nmalevich** [reply]
Hi @mbrandonw, thank you for detailed answer. It is really helpful
I have tried to replace Dependency(\\.dismiss) with Environment(\\.dismiss) and it just started working. So I'm grateful for your help

But I have faced with another problem. It pushes third view two times, after pop. I will send new project and you can look.
There is flow that you will need to do:
1->2->3->2->1->2->3(digits are screens)

I will be waiting your answer
[DismissErrorExample.zip](https://github.com/pointfreeco/swift-composable-architecture/files/14955927/DismissErrorExample.zip)

**Reply by nmalevich** [reply]
@mbrandonw Hi
Could I ask you have tried to produce this bug or you are going to?
I look forward to your answer, thank you

**Reply by mbrandonw** [reply]
Hi @nmalevich, I'm sorry but this is still navigation bugs in SwiftUI. I don't know of any workaround, other than to use `NavigationStack` with a binding, which we call "stack-based navigation." You can read our docs about how to do that, as well as the pros and cons between stack-based and tree-based navigation [here](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/navigation).

**Reply by nmalevich** [reply]
@mbrandonw Hi
Thank you for your answer. Ok, I suppose I will increase target to 16.0 and will use NavigationStack. I hope it will change the situation
Thanks a lot
